### PR TITLE
riscv: telink: change the nvs part to save ram.

### PR DESCRIPTION
**Pull request by haiwentelink**
- change nvs part and config.

### CODE REVIEW

File: include/zephyr/fs/nvs.h [hunk 1]
```diff
@@ -56,7 +66,12 @@ struct nvs_fs {
 	const struct device *flash_device;
 	const struct flash_parameters *flash_parameters;
 #if CONFIG_NVS_LOOKUP_CACHE
+	#ifdef CONFIG_NVS_SAVE_RAM_EN
```
Comment (liucoldstar) [inline include/zephyr/fs/nvs.h]: 建议改成这个名字：CONFIG_NVS_RAM_OPTIMIZATION

File: subsys/fs/nvs/Kconfig [hunk 2]
```diff
@@ -27,6 +27,23 @@ config NVS_LOOKUP_CACHE_SIZE
...
+	int "Non-volatile Storage lookup cache size"
+	default 128
+	range 1 65536
+	depends on NVS_LOOKUP_CACHE
+	help
+	  Number of entries in Non-volatile Storage lookup cache.
+	  It is recommended that it be a power of 2.
+
+
+config NVS_SAVE_RAM_EN
```
Comment (liucoldstar) [inline subsys/fs/nvs/Kconfig]: config NVS_RAM_OPTIMIZATION
    bool "Enable RAM optimization for NVS lookup cache"
    default n
    depends on NVS_LOOKUP_CACHE
    help
      Enable this option to optimize RAM usage for NVS by using a specialized RAM-based lookup cache.

File: subsys/fs/nvs/nvs.c [hunk 3]
```diff
@@ -40,6 +40,74 @@ static inline size_t nvs_lookup_cache_pos(uint16_t id)
 	return pos % CONFIG_NVS_LOOKUP_CACHE_SIZE;
 }
 
+#ifdef CONFIG_NVS_SAVE_RAM_EN
```
Comment (liucoldstar) [inline subsys/fs/nvs/nvs.c]: CONFIG_NVS_RAM_OPTIMIZATION

File: subsys/fs/nvs/nvs.c [hunk 4]
```diff
@@ -40,6 +40,74 @@ static inline size_t nvs_lookup_cache_pos(uint16_t id)
...
+			return cache_entry;
+		}
+		idx++;
+	}
+
+	return  (struct nvs_lookup *) NVS_LOOKUP_CACHE_NO_ADDR;
+}
+#endif
+
+#ifdef CONFIG_NVS_SAVE_RAM_EN
```
Comment (liucoldstar) [inline subsys/fs/nvs/nvs.c]: CONFIG_NVS_RAM_OPTIMIZATION

File: subsys/fs/nvs/nvs.c [hunk 5]
```diff
@@ -170,7 +239,22 @@ static int nvs_flash_ate_wrt(struct nvs_fs *fs, const struct nvs_ate *entry)
 #ifdef CONFIG_NVS_LOOKUP_CACHE
 	/* 0xFFFF is a special-purpose identifier. Exclude it from the cache */
 	if (entry->id != 0xFFFF) {
+		#ifdef CONFIG_NVS_SAVE_RAM_EN
```
Comment (liucoldstar) [inline subsys/fs/nvs/nvs.c]: CONFIG_NVS_RAM_OPTIMIZATION

File: subsys/fs/nvs/nvs.c [hunk 6]
```diff
@@ -630,11 +714,26 @@ static int nvs_gc(struct nvs_fs *fs)
 		}
 
 #ifdef CONFIG_NVS_LOOKUP_CACHE
+		#ifdef CONFIG_NVS_SAVE_RAM_EN
```
Comment (liucoldstar) [inline subsys/fs/nvs/nvs.c]: CONFIG_NVS_RAM_OPTIMIZATION

File: subsys/fs/nvs/nvs.c [hunk 7]
```diff
@@ -868,9 +967,15 @@ static int nvs_startup(struct nvs_fs *fs)
 		 * So, temporarily, we set the lookup cache to the end of the fs.
 		 * The cache will be rebuilt afterwards
 		 **/
+		#ifdef CONFIG_NVS_SAVE_RAM_EN
```
Comment (liucoldstar) [inline subsys/fs/nvs/nvs.c]: CONFIG_NVS_RAM_OPTIMIZATION

File: subsys/fs/nvs/nvs.c [hunk 8]
```diff
@@ -1035,11 +1140,23 @@ ssize_t nvs_write(struct nvs_fs *fs, uint16_t id, const void *data, size_t len)
 
 	/* find latest entry with same id */
 #ifdef CONFIG_NVS_LOOKUP_CACHE
+	#ifdef CONFIG_NVS_SAVE_RAM_EN
```
Comment (liucoldstar) [inline subsys/fs/nvs/nvs.c]: CONFIG_NVS_RAM_OPTIMIZATION

File: subsys/fs/nvs/Kconfig [hunk 9]
```diff
@@ -27,6 +27,23 @@ config NVS_LOOKUP_CACHE_SIZE
...
 
+config NVS_LOOKUP_CACHE_ARRAY_SIZE
+	int "Non-volatile Storage lookup cache size"
+	default 128
+	range 1 65536
+	depends on NVS_LOOKUP_CACHE
+	help
+	  Number of entries in Non-volatile Storage lookup cache.
+	  It is recommended that it be a power of 2.
+
```
Comment (liucoldstar) [inline subsys/fs/nvs/Kconfig:38]: 这里可以删一行

File: subsys/fs/nvs/nvs.c [hunk 10]
```diff
@@ -40,6 +40,74 @@ static inline size_t nvs_lookup_cache_pos(uint16_t id)
...
+	fs->lookup_idx = 0;
+	while (true) {
+		/* Make a copy of 'addr' as it will be advanced by nvs_pref_ate() */
+		ate_addr = addr;
+		rc = nvs_prev_ate(fs, &addr, &ate);
+
+		if (rc) {
+			return rc;
+		}
+
```
Comment (liucoldstar) [inline subsys/fs/nvs/nvs.c]: 这里也可以删一行空行

File: subsys/fs/nvs/nvs.c [hunk 11]
```diff
@@ -1035,11 +1140,23 @@ ssize_t nvs_write(struct nvs_fs *fs, uint16_t id, const void *data, size_t len)
 
 	/* find latest entry with same id */
 #ifdef CONFIG_NVS_LOOKUP_CACHE
+	#ifdef CONFIG_NVS_RAM_OPTIMIZATION
+	struct nvs_lookup *cache_entry = nvs_lookup_cache_location(fs, nvs_lookup_cache_pos(id));
+
+	if ((unsigned int)cache_entry == NVS_LOOKUP_CACHE_NO_ADDR) {
+
```
Comment (liucoldstar) [inline subsys/fs/nvs/nvs.c]: 删空行

File: subsys/fs/nvs/nvs.c [hunk 12]
```diff
@@ -1165,12 +1282,25 @@ ssize_t nvs_read_hist(struct nvs_fs *fs, uint16_t id, void *data, size_t len,
 	cnt_his = 0U;
 
 #ifdef CONFIG_NVS_LOOKUP_CACHE
+	#ifdef CONFIG_NVS_RAM_OPTIMIZATION
+	struct nvs_lookup *cache_entry = nvs_lookup_cache_location(fs, nvs_lookup_cache_pos(id));
+
+	if ((unsigned int)cache_entry == NVS_LOOKUP_CACHE_NO_ADDR) {
+
```
Comment (liucoldstar) [inline subsys/fs/nvs/nvs.c]: 删不必要的空行

**Review by andy-liu-telink — APPROVED** [review]

**Pull request merged**